Added core-image-bistro-full

DunderRoffe opened the pull request:
core-image-bistro-full is ment as an image that builds all the software provided by bistro.

This solves the problem with SoftwareContainer needing kernel-module-veth to work. Without veth SoftwareContainer can not start a container as LXC tries to create a virtual ethernet device linking a network bridge. However, it should still be possible to run SoftwareContainer if it is compiled without network support.

Gagi2k commented:
Although i see that this is useful to have, but why don't you add a PACKAGECONFIG option to softwarecontainer and depend on kernel-module-veth, if softwarecontainer is build with network support ?

tobsan commented:
I agree with @Gagi2k, and this image in itself does not solve the issue with kernel-module-veth. It seems depending on this does nothing if the kernel configuration is missing it, so one needs to have a kernel bbappend for it anyway (I hope I'm wrong, but that's how it looks).

If that is the case, this is a problem imho. Recipes should be able to simply DEPEND on other packages, and in the kernel case it would be nice if such a dependency would translate into the kernel config being changed automatically.

jonte commented:
@Gagi2k A PACKAGEOPTION will fail if the module is built in to the kernel, I think. I think adding it explicitly like this to the image recipe is better, since you should be able to rely on the platform always having this functionality. There are several other kernel features we simply rely on being there.

Maybe it would be better if softwarecontainer fails to build when the required kernel features are missing?

jonte commented on recipes-core/images/core-image-bistro-full.bb:
```diff
@@ -0,0 +1,11 @@
+#
+#   Copyright (C) 2015 Pelagicore AB
```
Please set this to 2016

erikboto commented:
Before we add an image that references both -ivi and -genivi images, what are the purpose of the -genivi image? It seems like it only includes a subset of GENIVI components, while -ivi brings in all P1 SC and P2 SC.

I don't think there's a way in Yocto to handle package dependencies towards kernel features. One can add the kernel-module-* package to RRECOMMENDS, but that doesn't work if the driver is in kernel as @jonte mentioned.

I think I would actually prefer to just leave it up to the "integrator" to make sure that the kernel is built with the features needed. Or maybe add RRECOMMENDS_${PN} += "kernel-module-veth", and then if that dependency fails during image creation the integrator can either add a BAD_RECOMMENDATION if the driver is in the kernel, or modify kernel config so the veth driver is built.

jonte commented:
@erikboto I think it would make sense to have some image which "just works" and has everything configured properly from start. If we don't do this in Bistro, then I think we need to add a new layer which contains these images.

I think some of the purpose of this merge request is to create a "just works" image, which makes sure to include the required kernel features, and all packages to do an effective demo of the things provided in Bistro.

We need to decide if we want to keep Bistro "clean" with no tweaking (such as deploying this kernel module), and put these tweaks in some other layer (meta-bistro-demo?), or if we want to include this in Bistro.

What do you guys think?

erikboto commented:
I agree that it would make sense to have an image that "just works". But that "just works" part can only be reliably achieved if it targets a specific target and comes with the BSP-modifications needed for that target. E.g. making sure that the kernel is configured according to softwarecontainers requirements, which I guess is a set of CGROUPS features, veth and maybe something more.

We could do the thing with multiple layers in the same git like many layers to, e.g. meta-ivi (http://git.yoctoproject.org/cgit/cgit.cgi/meta-ivi/tree/) and then do BSP modifications for certain machines in meta-bistro-bsp?

Also, I still wonder if the "require core-image-bistro-genivi.bb" add anything that "require core-image-bistro-ivi.bb" hasn't already pulled in. Do you know @DunderRoffe? Otherwise I think that line should be dropped and we should see if we can get rid of that image, it's just confusing to me since it just pulls in three of all GENIVI components.

socec commented:
@erikboto I also agree with removing "core-image-bistro-genivi". It installs persistence-client-library which is not included in P1 and P2 packagegroups directly, but one of the components there (node-state-manager) anyway has a DEPENDS on it. It also automatically includes core-image-bistro.

In general I agree with Erik that an image that "just works" can't be achieved easily in this case since it needs the proper kernel and BSP. I would vote for leaving this to the integrator with a RRECOMMENDS or something. I believe the idea of meta-bistro is to add functionality with recipes and sample configuration and sample images, but not be the top project layer for a project-specific image.

DunderRoffe commented:
@erikboto No I do not. I added it since it was a part of other images without much thought. So I guess it should be dropped.

I approach this as trying to build a reference version of Pelux on a dedicated target. From that point of view there meta-bistro would be the top layer. I guess one could add higher layers but that seems like an awfully fragmented solution.

The purpose of this pull-request is make sure that there were an image including a working SoftwareContainer. Since meta-bistro supplies the SoftwareContainer recipe I added it here. But I guess that this is better put at another level.

DunderRoffe commented:
After some discussions today this solution seems moot. Will remake this in another way, probably a separate layer.
